### PR TITLE
Improve remote translations

### PR DESCRIPTION
**Pull request by taitus**
## References
Related PR's: 
- [Translate user generated content #3359](https://github.com/consul/consul/pull/3359)
- [Update remote translations docs with create alert section #91](https://github.com/consul/docs/pull/91)

## Objectives
- Avoid curl requests to the application by requesting unavailable languages.
- Avoid curl requests to the application by requesting resources already translated 

## Visual Changes
None

## Notes
None

### CODE REVIEW

File: app/controllers/concerns/remotely_translatable.rb [hunk 1]
```diff
@@ -16,12 +16,16 @@ def remote_translation_for(resource)
     end
 
     def translation_empty?(resource)
-      resource.translations.where(locale: I18n.locale).empty?
+      resource.class.translates? && resource.translations.where(locale: I18n.locale).empty?
```
Comment (javierm) [inline app/controllers/concerns/remotely_translatable.rb]: Would it be possible to add a test case for this change?

File: app/models/remote_translation.rb [hunk 2]
```diff
@@ -17,4 +18,10 @@ def self.remote_translation_enqueued?(remote_translation)
           locale: remote_translation["locale"],
           error_message: nil).any?
   end
+
+  def already_translated_resource
+    if remote_translatable&.translations&.where(locale: locale).present?
+      errors.add(:locale, I18n.t("remote_translations.already_translated_resource"))
```
[the remark this answers is not in the document]
Comment (taitus) [reply app/models/remote_translation.rb]: Solved in the commit: cdb2937

File: app/controllers/concerns/remotely_translatable.rb [hunk 3]
```diff
@@ -2,7 +2,7 @@ module RemotelyTranslatable
   private
 
     def detect_remote_translations(*args)
-      return [] unless Setting["feature.remote_translations"].present?
+      return [] unless Setting["feature.remote_translations"].present? && api_key_has_been_set_in_secrets?
```
[the remark this answers is not in the document]
Comment (taitus) [reply app/controllers/concerns/remotely_translatable.rb]: Solved in commit: 086e38c (/spec/controllers/concerns/remotely_translatable_spec.rb)

File: app/controllers/concerns/remotely_translatable.rb (hunk 1 above)
Comment (taitus) [reply app/controllers/concerns/remotely_translatable.rb]: Solved in the commit: 18900da

File: spec/controllers/rack_attack_spec.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
+require "rails_helper"
+
+describe Rack::Attack, type: :request do
+  before do
+    RemoteTranslation.skip_callback(:create, :after, :enqueue_remote_translation)
+  end
+
+  after do
+    RemoteTranslation.set_callback(:create, :after, :enqueue_remote_translation)
+    avoid_test_overlaps_in_cache
```
[the remark this answers is not in the document]
Comment (taitus) [reply spec/controllers/rack_attack_spec.rb]: Solved in the new PR created to add rack-attack gem: #3928

File: spec/features/stats_spec.rb [hunk 5]
```diff
@@ -1,6 +1,10 @@
 require "rails_helper"
 
 describe "Stats" do
+  before do
```
[the remark this answers is not in the document]
Comment (taitus) [reply spec/features/stats_spec.rb]: I think I like the "after" option better.
Solved in the new PR created to add rack-attack gem: #3928

File: Gemfile [hunk 6]
```diff
@@ -41,6 +41,7 @@ gem "paranoia", "~> 2.4.2"
 gem "pg", "~> 0.21.0"
 gem "pg_search", "~> 2.0.1"
 gem "puma", "~> 4.3.1"
+gem "rack-attack"
```
[the remark this answers is not in the document]
Comment (taitus) [reply Gemfile]: Solved in the new PR created to add rack-attack gem: #3928

File: spec/controllers/concerns/remotely_translatable_spec.rb [hunk 7]
```diff
@@ -48,11 +49,29 @@
...
+    end
+
     it "When defined in current locale should not detect remote_translations" do
       proposal = create(:proposal)
       comment = create(:comment, commentable: proposal)
 
       expect(detect_remote_translations([proposal, comment])).to eq []
     end
+
+    it "When resource class has not translates should not detect remote_translations" do
```
Comment (javierm) [inline spec/controllers/concerns/remotely_translatable_spec.rb]: "has not translates" seems to be a typo :thinking:. How about "is not translatable"?

File: app/models/remote_translation.rb [hunk 8]
```diff
@@ -17,4 +18,10 @@ def self.remote_translation_enqueued?(remote_translation)
           locale: remote_translation["locale"],
           error_message: nil).any?
   end
+
+  def already_translated_resource
+    if remote_translatable&.translations&.where(locale: locale).present?
+      errors.add(:locale, I18n.t("activerecord.errors.models.remote_translation.attributes.locale.already_translated"))
```
[the remark this answers is not in the document]
Comment (javierm) [reply app/models/remote_translation.rb]: Oh, sorry, I didn't explain it properly last time :sweat_smile:. What I meant was we could simplify the code and write:

```ruby
errors.add(:locale, :already_translated)
```

Then Rails automatically gets finds the translation key.

**Pull request merged**